src/ERC20Reserve.sol

```diff
@@ -5,54 +5,49 @@ import {Ownable} from "openzeppelin-contracts/access/Ownable.sol";
 import {IERC20} from "openzeppelin-contracts/token/ERC20/IERC20.sol";
 
 interface IERC20Reserve {
-    function erc20() external view returns (IERC20);
+    function withdrawable(IERC20 token) external view returns (uint256 amt);
 
-    function withdraw(uint256 amt) external;
+    function withdraw(IERC20 token, uint256 amt) external;
 
-    function deposit(uint256 amt) external;
+    function deposit(IERC20 token, uint256 amt) external;
 }
 
 contract ERC20Reserve is IERC20Reserve, Ownable {
-    IERC20 public immutable override erc20;
     address public user;
-    uint256 public balance;
+    mapping(IERC20 => uint256) public override withdrawable;
 
     event Withdrawn(address to, uint256 amt);
     event Deposited(address from, uint256 amt);
     event ForceWithdrawn(address to, uint256 amt);
     event UserSet(address oldUser, address newUser);
 
-    constructor(
-        IERC20 _erc20,
-        address owner,
-        address _user
-    ) {
-        erc20 = _erc20;
+    constructor(address owner, address _user) {
         setUser(_user);
         transferOwnership(owner);
     }
 
     modifier onlyUser() {
-        require(_msgSender() == user, "Reserve: caller is not the user");
+        require(msg.sender == user, "Reserve: caller is not the user");
         _;
     }
 
-    function withdraw(uint256 amt) public override onlyUser {
+    function withdraw(IERC20 token, uint256 amt) public override onlyUser {
+        uint256 balance = withdrawable[token];
         require(balance >= amt, "Reserve: withdrawal over balance");
-        balance -= amt;
-        emit Withdrawn(_msgSender(), amt);
-        require(erc20.transfer(_msgSender(), amt), "Reserve: transfer failed");
+        withdrawable[token] = balance - amt;
+        require(token.transfer(msg.sender, amt), "Reserve: transfer failed");
+        emit Withdrawn(msg.sender, amt);
     }
 
-    function deposit(uint256 amt) public override onlyUser {
-        balance += amt;
-        emit Deposited(_msgSender(), amt);
-        require(erc20.transferFrom(_msgSender(), address(this), amt), "Reserve: transfer failed");
+    function deposit(IERC20 token, uint256 amt) public override onlyUser {
+        withdrawable[token] += amt;
+        require(token.transferFrom(msg.sender, address(this), amt), "Reserve: transfer failed");
+        emit Deposited(msg.sender, amt);
     }
 
-    function forceWithdraw(uint256 amt) public onlyOwner {
-        emit ForceWithdrawn(_msgSender(), amt);
-        require(erc20.transfer(_msgSender(), amt), "Reserve: transfer failed");
+    function forceWithdraw(IERC20 token, uint256 amt) public onlyOwner {
+        require(token.transfer(msg.sender, amt), "Reserve: transfer failed");
```

Comment:
Mhm, I think we could reduce the withdrawable here but allow to withdraw more than the balance.
```solidity
if(balance <= amt) {
 withdrawable[token] = 0;
} else {
 withdrawable[token] = balance - amt;
}
```

Reply:
One of the use cases for forced withdrawal is to shave off tokens which are in the reserve but will never be withdrawn, e.g. the DSR-generating income. If we reduced `withdrawable`, the drips hub could get stuck.

Forced withdrawal is a counterpart of ERC-20 transfers, they too allow changing the actual balance of the reserve without touching `withdrawable`.

Reply:
Ah, fair point to keep it like that. Maybe we could call the function `harvest` which only allows to get the difference between `erc20.balanceOf` and the stored `balance`. 

If you want all of it, you could still make yourself a regular user.

Reply:
Yeah, adding `harvest` could make sense :thinking:

Pros:
- More difficult for `forceWithdraw` to make the reserve go below `withdrawable` by mistake
Cons:
- More complex API with more logic and weirder, harder to use functions

`harvest` can be emulated using `forceWithdraw` as:
```solidity
function harvest(uint256 amt) {
    require(amt < token.balanceOf(address(reserve)) - reserve.withdrawable(token));
    reserve.forceWithdraw(amt);
}